kvstore/pebble/kvstore.go

```diff
@@ -0,0 +1,163 @@
+	db, err := pebble.Open(path, opts)
+	if err != nil {
+		return nil, errors.Join(ErrPebbleOpeningStore, err)
+	}
+	store.db = db
+	return store, nil
+}
+
+// Set stores a key-value pair in the database.
+func (store *pebbleKVStore) Set(key, value []byte) error {
+	if key == nil {
+		return ErrPebbleUnableToSetValue
+	}
+	err := store.db.Set(key, value, pebble.Sync)
+	if err != nil {
+		return errors.Join(ErrPebbleUnableToSetValue, err)
+	}
+	return nil
+}
+
+// Get retrieves the value associated with the given key.
+func (store *pebbleKVStore) Get(key []byte) ([]byte, error) {
+	if key == nil {
+		return nil, ErrPebbleUnableToGetValue
+	}
+	value, closer, err := store.db.Get(key)
+	if err != nil {
+		if err == pebble.ErrNotFound {
+			return nil, ErrPebbleUnableToGetValue
+		}
+		return nil, errors.Join(ErrPebbleUnableToGetValue, err)
+	}
+	defer closer.Close()
+	return append([]byte{}, value...), nil
+}
+
+// Delete removes the key-value pair associated with the given key.
+func (store *pebbleKVStore) Delete(key []byte) error {
+	if key == nil {
+		return ErrPebbleUnableToDeleteValue
+	}
+	err := store.db.Delete(key, pebble.Sync)
+	if err != nil {
+		return errors.Join(ErrPebbleUnableToDeleteValue, err)
+	}
+	return nil
+}
+
+// GetAll retrieves all key-value pairs with keys starting with the given prefix.
+// If descending is true, it returns the results in reverse lexicographical order.
+func (store *pebbleKVStore) GetAll(prefix []byte, descending bool) (keys, values [][]byte, err error) {
+	iter, _ := store.db.NewIter(&pebble.IterOptions{
+		LowerBound: prefix,
+		UpperBound: prefixEndBytes(prefix),
+	})
+	defer iter.Close()
+
+	if descending {
+		for valid := iter.Last(); valid; valid = iter.Prev() {
+			keys = append(keys, append([]byte{}, iter.Key()...))
+			values = append(values, append([]byte{}, iter.Value()...))
+		}
+	} else {
+		for iter.First(); iter.Valid(); iter.Next() {
+			keys = append(keys, append([]byte{}, iter.Key()...))
+			values = append(values, append([]byte{}, iter.Value()...))
+		}
+	}
+
+	if err := iter.Error(); err != nil {
+		return nil, nil, errors.Join(ErrPebbleIteratingStore, err)
+	}
+
+	return keys, values, nil
+}
+
+// Exists checks if a key exists in the store and has a non-empty value.
+func (store *pebbleKVStore) Exists(key []byte) (bool, error) {
+	value, closer, err := store.db.Get(key)
+	if err == pebble.ErrNotFound {
+		return false, nil
+	}
+	if err != nil {
+		return false, errors.Join(ErrPebbleUnableToGetValue, err)
+	}
+	defer closer.Close()
+	return len(value) > 0, nil
+}
+
+// ClearAll removes all key-value pairs from the store.
+func (store *pebbleKVStore) ClearAll() error {
+	iter, _ := store.db.NewIter(nil)
+	defer iter.Close()
+	for iter.First(); iter.Valid(); iter.Next() {
+		if err := store.db.Delete(iter.Key(), pebble.Sync); err != nil {
+			return errors.Join(ErrPebbleClearingStore, err)
+		}
+	}
+	if err := iter.Error(); err != nil {
+		return errors.Join(ErrPebbleClearingStore, err)
+	}
```

Comment:
Random thought, would it be faster to just zero the db file?

Reply:
We shouldn't zero/rm the file(s) if the database is still open.

In RelayMiner, I proposed not to delete the keys anymore (making this function obsolete) because we remove the database itself in the next step. 

So you're correct - we don't need to delete the keys.

Reply:
I think we need to remember this being a separate repo means it shouldn't solely focus on Shanon so keep it in - for the interface alone mark it with a potential todo to see if theres a faster way to clear it.

Reply:
There are ways to optimize. For example we can use Batch + DeleteRange. Will add a comment.